### DISCUSSION TITLE
Private, server-side only API calls - best practices?

### DISCUSSION THREAD

**Discussion by alasdairbarton** [nuxt/nuxt#16524] [Questions] [answered]
I'm messing around with nuxt 3 and I'm trying to understand the best way to make this work.

I've got an API, let's call it `api.server-a.com`, and I'm authenticating requests using a static API key (passed in through runtime environment variables), which is designed to **only** be used on the backend - I don't want to reveal it to end-users, but I still want to serve up data on the page.

I'd like to be able to make the API call on the server, generate the HTML, and then pass it to the client without any keys being exposed, or the client having to make any API calls.

I've tried this using refs and conditional logic using `process.server` and `process.client` to distinguish between the two:
```js
<template>
  <h1>{{ apiData.some_value }}</h1>
</template>
<script setup>
const apiData = ref()

if (process.server) {
  const { apiToken, apiEndpoint } = useRuntimeConfig();

  apiData.value = (
    (
      await useFetch(`${apiEndpoint}/Main_Page`, {
        method: 'GET',
        headers: { Authorization: `Bearer ${apiToken}` },
      })
    ).data.value as MainPageData
  ).data;
}
</script>
```
but this leads to hydration mismatches and errors in the client console, which I'd ideally like to avoid.

As far as I've read from the docs, I should be able to use `useState('apiData', () => <call api here>)` to persist the state through hydration, but I've not been able to try this as useState doesn't accept asynchronous functions.

Sorry for the long-winded post, if anyone's got any ideas I'd really appreciate it.

**Comment by Diizzayy** [comment, marked as answer]
@thebartyparty Solely executing your request within in `process.server` wouldn't be suffice to attain the experience that you're looking for as doing so would cripple client side navigation ( a user navigating to from `page a` to `page b` ) as the client wouldn't be aware of the request logic.

As for the hydration mismatch

> but this leads to hydration mismatches and errors in the client console, which I'd ideally like to avoid.

This occurs because, on initial load (server server) it is aware of `apiData` as well as the update to it's value by your api call, the caveat now is that, on the client side `apiData` is essentially null. So upon loading the page, you may encounter a brief flash where the requested data is initially hydrated then updates to a blank state once the client side scripts have completed loading.

Nuxt 3 doesn't currently provide a way for data to only be loaded on server side.

However though, you can create a [Nuxt Server Route](http://v3.nuxtjs.org/guide/features/server-routes#server-routes-1) where your authenticated requests would be made on the server, this route can then be used in your pages to proxy the private.

`nuxt-app/server/api/data.ts`
```ts
export default defineEventHandler(async (event) => {
  const { apiToken, apiEndpoint } = useRuntimeConfig();

  return $fetch(`${apiEndpoint}/Main_Page`, {
    method: "GET",
    headers: { Authorization: `Bearer ${apiToken}` },
  });
});
```

within your page
```vue
<template>
    <div>
        <p>
            data: {{ data }}
        </p>
    </div>
</template>

<script setup>
const { data } = await useFetch('/api/data', {
    // initialCache: true // this defaults true, it caches the first requests and simply returns the cached data for subsequent requests
})
</script>
```

**Reply by alasdairbarton** [reply]
Thanks! That's really helpful, greatly appreciate the in-depth response. I'll implement it for now - do you think this will be possible when full static arrives? I guess it just depends on whether full static means 'without hydration' or something else. I'm not really using this for the most dynamic thing ever - the less work done on the edge, the better.

**Reply by leccyril** [reply]
the problem with this method is the status returned by the client call is almost always status 200....

for example my call is returned specific status that i need to get on client side, but to make it work i have to process in server API, then put in response status.. that is not clean

There is no possibility to have like the middleware executed on server side and add for example header token etc (like an interceptor but only executed on server side to not exposes keys )? i try with server middleware but only called if i call server api, if from client side i fetch external application, middleware is not call
